### DISCUSSION TITLE
FVL 4.0 and the plugin system

### DISCUSSION THREAD

**Discussion by marina-mosti** [formvuelate/formvuelate#269] [Ideas]
We've gotten to a difficult but exciting point in the life of FormVueLate.
Where we have realized the plugin system is very powerful but flawed because of a couple reasons:

- FVL attempts to parse the schema on a per SchemaForm level, this causes a problem with plugins like Lookup that want to preparse the whole schema. 
- The SchemaForm parsing happens first, and then the plugin's parse which creates a difficult to solve race condition

Several attempts at fixing this have helped us understand that the problem can no longer be addressed without resorting to unmaintanable spaghetti code or the addition of a lot of bloat API, which brings us to the following conlusion.

The point in time has come for FVL to bring in yet another refactor/rewrite that will mostly focus on:

1. Moving Lookup to be part of the core FVL lib. 
2. Rewriting the way that schema parsing happens to better serve the plugin system without affecting nested schemas
3. Rethinking if necessary the way that we allow plugins to interact with FVL itself

Some breaking changes are to be expected at this point, mainly with Lookup getting absorved into core.
As a first look, I think we should approach it the following way:

1. Users will remove the `LookupPlugin` import fully
2.  The params that are currently passed at lookup, which is an object with `mapProps` and `mapComponents` as well as the second param to inject the local components will be moved into `useSchemaForm`
3. All necessary manipulation hereforth of the schema by core lib and internal lookup should be done at `useSchemaForm` composable since thats the primary entry point for the users' model.

```js
import { useSchema } from 'formvuelate'

...
const schema = { mySchema }
const initialModelState = {}

// Breaking change here on the params where the initial single param gets mixed into the config object as initialModelState
// The config obj also extends flexibility for future proofing
const { model } = useSchemaForm({
  initialModelState,
  mapProps: {},
  mapComponents: {},
  localComponents: [1,2,3]
})
```

**Comment by logaretm** [comment]
This is a good overview of what we want to do, I have a few thoughts on the proposed API

The `mapComponents` and `localComponents` could serve similar purposes, for example we can do this:

```js
const { model } = useSchemaForm({
  initialModelState,
  mapProps: {},
  components: {
     'text': InputText, // Defines a string-to-component mapping
     InputCheckbox, // Defines a component, since the string and component name matches
     'select': (el) => el.multiple ? InputMultiSelect : InputSelect, // Dynamic assignment based on other attributes in schema
  },
})
```

And it kinda translates well with Vue's component definitions.

What do you think, have I missed some cases?

**Reply by marina-mosti** [reply]
I like this idea a lot @logaretm   I can't think right now of any edge cases or problems at face value but we won't know until we implement. 
I like the look of the api though it seems clearer
